File: backend/routers/images.py
"""
FILE: backend/routers/images.py
Image generation via Cloudflare Workers AI (Stable Diffusion).
"""

from fastapi import APIRouter, Depends
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select

from database import get_db
from models import GeneratedFile
from services.image_gen import generate_image

router = APIRouter()


class ImageRequest(BaseModel):
    prompt: str
    negative_prompt: str = "blurry, low quality, distorted, ugly, watermark"
    width: int = 1024
    height: int = 1024
    steps: int = 20


@router.post("/generate")
async def generate(req: ImageRequest, db: AsyncSession = Depends(get_db)):
    """Generate an image from a text prompt."""
    result = await generate_image(
        prompt=req.prompt,
        negative_prompt=req.negative_prompt,
        width=req.width,
        height=req.height,
        steps=req.steps,
    )

    # Save record
    gf = GeneratedFile(
        filename=result["filename"],
        file_type="png",
        description=req.prompt,
        url_path=result["url"],
    )
    db.add(gf)
    await db.commit()

    return result


@router.get("/")
async def list_images(db: AsyncSession = Depends(get_db)):
    result = await db.execute(
        select(GeneratedFile)
        .where(GeneratedFile.file_type == "png")
        .order_by(GeneratedFile.created_at.desc())
        .limit(50)
    )
    images = result.scalars().all()
    return [
        {
            "id": img.id,
            "url": img.url_path,
            "description": img.description,
            "created_at": img.created_at.isoformat(),
        }
        for img in images
    ]

File: cf-backend/src/routes/images.ts
// FILE: cf-backend/src/routes/images.ts
import { Hono } from "hono";
import { Env } from "../types";
import { generateImage } from "../services/ai";

export const imagesRouter = new Hono<{ Bindings: Env }>();

imagesRouter.post("/generate", async (c) => {
  const {
    prompt,
    negative_prompt = "blurry, low quality, distorted, ugly, watermark",
    steps = 20,
  } = await c.req.json<{ prompt: string; negative_prompt?: string; steps?: number }>();

  if (!prompt) return c.json({ error: "prompt required" }, 400);

  const imageBuffer = await generateImage(c.env, prompt, steps);
  const bytes = new Uint8Array(imageBuffer);
let binary = "";
for (let i = 0; i < bytes.byteLength; i++) {
  binary += String.fromCharCode(bytes[i]);
}
const base64 = btoa(binary);
  const url = `data:image/png;base64,${base64}`;

  return c.json({ url, prompt, width: 1024, height: 1024 });
});

imagesRouter.get("/", async (c) => {
  return c.json([]);
});
